# Sample reports

---
output:
  html_document:
   highlight: tango
---

```{r load-packages, include=FALSE}
library(data.table)
library(jsonlite)

```



```{r, include=FALSE}
config_json = as.data.table(fromJSON(snakemake@params$config))

```
Overview report of workflow done for sample `r snakemake@params$sample_name `.




## Report files:
`r paste0("* [Main QC summary - (MultiQC)](../../",snakemake@input$multiqc,")  \\")`
`r if(unique(snakemake@params$paired) == "PE"){paste0("* [Detailed raw fastq QC - (FastQC)](../../",snakemake@input$raw_fastq_R1_report,")  \\")} `
`r if(unique(snakemake@params$paired) == "PE"){paste0("* [Detailed raw fastq QC - (FastQC)](../../",snakemake@input$raw_fastq_R2_report,")  \\")} `
`r if(unique(snakemake@params$paired) == "SE"){paste0("* [Detailed raw fastq QC - (FastQC)](../../",snakemake@input$raw_fastq_SE_report,")  \\")} `
`r if(unique(config_json$qc_qualimap_DNA) == TRUE){paste0("* [Detailed mapping QC - Qualimap bamqc](../../",snakemake@input$qc_qualimap_DNA,")  \\")} `
`r if(unique(config_json$qc_qualimap_RNA) == TRUE){paste0("* [RNA-Seq metrics - Qualimap rnaseq](../../",snakemake@input$qc_qualimap_RNA,")  \\")} `
`r if(unique(config_json$qc_picard_RNA) == TRUE){paste0("* [RNA-Seq metrics - Picard](../../",snakemake@input$qc_picard_RNA,")  \\")} `
`r if(unique(config_json$qc_fastq_screen_RNA) == TRUE){paste0("* [RNA-Seq metrics - Fastq_screen](../../",snakemake@input$qc_fastq_screen_RNA,")  \\")} `
`r if(unique(config_json$qc_biotypes_RNA) == TRUE){paste0("* [RNA-Seq metrics - Biotypes](../../",snakemake@input$qc_biotypes_RNA,")  \\")} `
`r if(unique(config_json$qc_RSeQC_RNA) == TRUE){paste0("* [RNA-Seq metrics - RSeQC RNA](../../",snakemake@input$qc_RSeQC_RNA,")  \\")} `
`r if(unique(config_json$featureCount_exon) == TRUE){paste0("* [RNA-Seq metrics - featureCount exon](../../",snakemake@input$featureCount_exon,")  \\")} `
`r if(unique(config_json$featureCount_gene) == TRUE){paste0("* [RNA-Seq metrics - featureCount gene](../../",snakemake@input$featureCount_gene,")  \\")} `
`r if(unique(config_json$featureCount_transcript) == TRUE){paste0("* [RNA-Seq metrics - featureCount transcript](../../",snakemake@input$featureCount_transcript,")  \\")} `
`r if(unique(config_json$featureCount_3pUTR) == TRUE){paste0("* [RNA-Seq metrics - featureCount 3 prime UTR](../../",snakemake@input$featureCount_3pUTR,")  \\")} `
`r if(unique(config_json$featureCount_5pUTR) == TRUE){paste0("* [RNA-Seq metrics - featureCount 5 prime UTR](../../",snakemake@input$featureCount_5pUTR,")  \\")} `
`r if(unique(config_json$RSEM) == TRUE){paste0("* [RNA-Seq metrics - RSEM](../../",snakemake@input$RSEM,")  \\")} `
`r if(unique(config_json$chip_extra_qc) == TRUE){paste0("* [ChIP-Seq metrics - Phantompeakqualtools (no duplicates)](../",snakemake@input$phantompeak, ") \\")} `
`r if(unique(config_json$chip_extra_qc) == TRUE){paste0("* [ChIP-Seq metrics - Phantompeakqualtools (with duplicates)](../",snakemake@input$phantompeak_dups, ") \\")} `

----------------------------------------------------------

Return to [start page](../final_alignment_report.html)

----------------------------------------------------------